text-input.go

```diff
@@ -0,0 +1,289 @@
+	p.channel.HandleFuncSync("TextInput.clearClient", p.handleClearClient)
+	p.channel.HandleFuncSync("TextInput.setEditingState", p.handleSetEditingState)
+	p.channel.HandleFunc("TextInput.show", func(_ interface{}) (interface{}, error) {
+		if p.virtualKeyboardShow != nil {
+			p.virtualKeyboardShow()
+		}
+		return nil, nil
+	})
+	p.channel.HandleFunc("TextInput.hide", func(_ interface{}) (interface{}, error) {
+		if p.virtualKeyboardHide != nil {
+			p.virtualKeyboardHide()
+		}
+		return nil, nil
+	})
+	// Ignored: This information is used by the flutter Web Engine
+	p.channel.HandleFuncSync("TextInput.setStyle", func(_ interface{}) (interface{}, error) { return nil, nil })
+	// Ignored: This information is used by the flutter Web Engine
+	p.channel.HandleFuncSync("TextInput.setEditableSizeAndTransform", func(_ interface{}) (interface{}, error) { return nil, nil })
+	// Ignored: This information is used by flutter on Android, iOS and web
+	p.channel.HandleFuncSync("TextInput.requestAutofill", func(_ interface{}) (interface{}, error) { return nil, nil })
+
+	return nil
+}
+
+func (p *textinputPlugin) handleSetClient(arguments interface{}) (reply interface{}, err error) {
+	args := []json.RawMessage{}
+	err = json.Unmarshal(arguments.(json.RawMessage), &args)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode json arguments for handleSetClient")
+	}
+
+	err = json.Unmarshal(args[0], &p.clientID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode clientID for handleSetClient")
+	}
+
+	err = json.Unmarshal(args[1], &p.clientConf)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode clientConf for handleSetClient")
+	}
+
+	return nil, nil
+}
+
+func (p *textinputPlugin) handleClearClient(arguments interface{}) (reply interface{}, err error) {
+	p.clientID = 0
+	return nil, nil
+}
+
+func (p *textinputPlugin) handleSetEditingState(arguments interface{}) (reply interface{}, err error) {
+	if p.clientID == 0 {
+		return nil, errors.New("cannot set editing state when no client is selected")
+	}
+
+	err = json.Unmarshal(arguments.(json.RawMessage), &p.ed)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode json arguments for handleSetEditingState")
+	}
+
+	p.ed.utf16Text = utf16.Encode([]rune(p.ed.Text))
+	utf16TextLen := len(p.ed.utf16Text)
+
+	// sometimes flutter sends invalid cursor position
+	if p.ed.SelectionBase < 0 ||
+		p.ed.SelectionExtent < 0 ||
+		p.ed.SelectionBase > utf16TextLen ||
+		p.ed.SelectionExtent > utf16TextLen {
+		// request a new EditingState
+		err := p.channel.InvokeMethod("TextInputClient.requestExistingInputState", nil)
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (p *textinputPlugin) glfwCharCallback(w *glfw.Window, char rune) {
+	if p.clientID == 0 {
+		return
+	}
+	// Opinionated: If a flutter dev uses TextCapitalization.characters
+	//              in a TextField, that means he wants only to receive
+	//              uppercase characters.
```

Comment:
Interesting. In some cases it may be possible to switch back from the uppercased keyboard to the lowercased one... Although indeed that may not be the intention of the developer. We'll have to see how this lands with go-flutter users. For now I think this is the best approach yes.